eclipse-theia/theia
# Integrate Memory Inspector in Theia repository

#### What it does


This PR transfers the Memory Inspector functionality from [the Theia CPP Extensions](https://github.com/eclipse-theia/theia-cpp-extensions) to the Theia repository. Relative to the current state of that repository, this PR makes some minor changes necessary to operate with Theia@1.27.0 and transfers some functionality that is not fully determined by the Debug Adapter Protocol into the `MemoryProviderService`, where it can be more easily adjusted for adapters with specific known properties.

#### How to test


1. With the CDT-GDB adapter, the instructions from [this PR](https://github.com/eclipse-theia/theia-cpp-extensions/pull/119) still apply
2. To demonstrate the functionality with another debug adapter: (less convenient than GDB, because it supports fewer reference options)
3. Create and open a new Rust project or use [this one](https://github.com/colin-grant-work/demo-extensions/tree/main/struct_memory)
4. In your workspace, make sure that the preference `lldb.rpcServer` is set to `null` to work around #11392
5. Install `Rust Analyzer` and `CodeLLDB`
6. (If using the example project linked above) open `main.rs` and put a breakpoint on line 12.
7. Use the code lenses provided by Rust Analyzer to start a debug session from the testson lines 16-25.
8. When you hit the breakpoint, the terminal in which the program is running will have printed two addresses.
9. Open the memory inspector from the right panel and input one of the addresses into the address selector and hit `Go`. The memory location should load, and it should match the data of the struct whose address you entered.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
